File: java/src/main/java/org/twyszomirski/aoc/solutions/Day_8.java
package org.twyszomirski.aoc.solutions;


import org.twyszomirski.aoc.utils.Pair;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import static java.io.IO.println;

public class Day_8 {

    public static void solve() throws IOException {
        var lines = Files.readAllLines(Paths.get("src/main/resources/input_day_8.txt"));
        println("======== Day 8 ===========");
        part_1(lines);
        part_2(lines);
    }

    private static void part_1(List<String> lines) {
        var data = readData(lines);
        var grouped = data.stream().filter(AntennaPoint::notEmpty)
                .collect(Collectors.groupingBy(AntennaPoint::symbol));

        var count = grouped.values().stream()
                .flatMap(antenna -> calculatePairs(antenna).stream())
                .flatMap(pairs ->
                        calculateAntinodes(pairs, lines.size(), lines.getFirst().length()).stream())
                .collect(Collectors.toSet()).size();
        println(count);
    }

    private static void part_2(List<String> lines) {
        var data = readData(lines);
        var grouped = data.stream().filter(AntennaPoint::notEmpty)
                .collect(Collectors.groupingBy(AntennaPoint::symbol));

        var antinodes = grouped.values().stream()
                .flatMap(antenna -> calculatePairs(antenna).stream())
                .flatMap(pairs ->
                        calcInfiniteAntiNodes(pairs, lines.size(), lines.getFirst().length()).stream())
                .collect(Collectors.toSet());

        var antinodesOfPairs = grouped.values().stream()
                .flatMap(antenna -> calculatePairs(antenna).stream())
                .flatMap(it -> Stream.of(it.left(), it.right()))
                .map(it -> new AntennaPoint(it.row, it.column, "#")).collect(Collectors.toSet());

        antinodes.addAll(antinodesOfPairs);

        println(antinodes.size());
    }

    private static List<AntennaPoint> readData(List<String> lines) {
        return IntStream.range(0, lines.size()).mapToObj(row -> {
            var split = lines.get(row).split("");
            return IntStream.range(0, split.length).mapToObj(col ->
                    new AntennaPoint(row, col, split[col])).toList();
        }).flatMap(Collection::stream).toList();
    }

    private static List<Pair<AntennaPoint>> calculatePairs(List<AntennaPoint> points) {
        if (points.size() < 2) return Collections.emptyList();
        List<Pair<AntennaPoint>> current = points.stream().skip(1).map(p -> new Pair<>(points.getFirst(), p)).toList();
        var result = new ArrayList<>(current);
        result.addAll(calculatePairs(points.stream().skip(1).toList()));
        return result;
    }

    private static List<AntennaPoint> calculateAntinodes(Pair<AntennaPoint> pair, Integer numOfRows, Integer numOfColumns) {
        return Stream.of(calcUpperAntinode(pair), calcLowerAntinode(pair))
                .filter(it -> it.inMatrix(numOfRows, numOfColumns))
                .toList();
    }

    private static List<AntennaPoint> calcInfiniteAntiNodes(Pair<AntennaPoint> pair, Integer numOfRows, Integer numOfColumns) {
        var upper = new ArrayList<AntennaPoint>();
        var nextPair = pair;
        do {
            upper.add(calcUpperAntinode(nextPair));
            nextPair = new Pair(upper.getLast(), nextPair.left());
        } while (upper.getLast().inMatrix(numOfRows, numOfColumns));

        var lower = new ArrayList<AntennaPoint>();
        nextPair = pair;
        do {
            lower.add(calcLowerAntinode(nextPair));
            nextPair = new Pair(nextPair.right(), lower.getLast());
        } while (lower.getLast().inMatrix(numOfRows, numOfColumns));


        var res = new ArrayList<>(upper);
        res.addAll(lower);
        return res.stream().filter(it -> it.inMatrix(numOfRows, numOfColumns)).toList();
    }

    private static AntennaPoint calcUpperAntinode(Pair<AntennaPoint> antennas) {
        var differenceRow = (antennas.left().row - antennas.right().row);
        var differenceCol = (antennas.left().column - antennas.right().column);
        return new AntennaPoint(antennas.left().row + differenceRow, antennas.left().column + differenceCol, "#");
    }

    private static AntennaPoint calcLowerAntinode(Pair<AntennaPoint> antennas) {
        var differenceRow = (antennas.left().row - antennas.right().row);
        var differenceCol = (antennas.left().column - antennas.right().column);
        return new AntennaPoint(antennas.right().row - differenceRow, antennas.right().column - differenceCol, "#");
    }

    record AntennaPoint(Integer row, Integer column, String symbol) {
        private boolean notEmpty() {
            return !".".equals(symbol);
        }

        private boolean inMatrix(Integer numOfRows, Integer numOfColumns) {
            return row >= 0 && column >= 0 && row < numOfRows && column < numOfColumns;
        }
    }


}

File: kotlin/src/main/kotlin/org/twyszomirski/aoc/solutions/Day_8.kt
package org.twyszomirski.aoc.solutions

import java.io.File


class Day_8 {

    fun solve() {
        val lines = File("src/main/resources/input_day_8.txt").readLines()
        println("======== Day 8 ===========")
        part1(lines)
        part2(lines)
    }

    fun part1(input: List<String>) {
        val data = readData(input)
        val numCols = data.first().size
        val numRows = data.size

        val bySymbol = data.flatten().filter { !it.antenna.equals(".") }.groupBy { it.antenna }
        val pairs = bySymbol.mapValues { producePairs(it.value) }
        val antinodes = pairs.mapValues {
            it.value.flatMap { x -> calcAntiNodes(x) }
        }

        val calc = antinodes.values.flatten().toSet()
            .count { it.isInMatrix(numRows, numCols) }

        println(calc)
    }

    fun part2(input: List<String>) {
        val data = readData(input)
        val numCols = data.first().size
        val numRows = data.size

        val bySymbol = data.flatten().filter { !it.antenna.equals(".") }.groupBy { it.antenna }
        val pairs = bySymbol.mapValues { producePairs(it.value) }
        val antinodes = pairs.mapValues {
            it.value.flatMap { x -> calcInfiniteAntiNodes(x, numRows, numCols) }
        }

        val antinodesOfPairs = pairs.values.flatten()
            .flatMap { listOf(Point(it.first.row, it.first.column, "#"), Point(it.second.row, it.second.column, "#")) }
            .toSet()

        val calc = (antinodes.values.flatten().toSet() + antinodesOfPairs)
            .count { it.isInMatrix(numRows, numCols) }

        println(calc)
    }

    fun producePairs(points: List<Point>): List<Pair<Point, Point>> {
        if (points.size < 2) {
            return emptyList()
        }
        val first = points.first()
        val rest = points.subList(1, points.size)
        val currentPairs = rest.map { Pair(first, it) }
        return currentPairs + producePairs(rest)
    }

    fun calcAntiNodes(antennas: Pair<Point, Point>): List<Point> {
        return listOf(calcUpperAntinode(antennas), calcLowerAntinode(antennas))
    }

    fun calcInfiniteAntiNodes(antennas: Pair<Point, Point>, columns: Int, rows: Int): List<Point> {
        val upper = mutableListOf<Point>()
        var nextPair = antennas
        do {
            upper.add(calcUpperAntinode(nextPair))
            nextPair = Pair(upper.last(), nextPair.first)
        } while (upper.last().isInMatrix(rows, columns))

        val lower = mutableListOf<Point>()
        nextPair = antennas
        do {
            lower.add(calcLowerAntinode(nextPair))
            nextPair = Pair(nextPair.second, lower.last())
        } while (lower.last().isInMatrix(rows, columns))

        return upper + lower
    }

    fun calcUpperAntinode(antennas: Pair<Point, Point>): Point {
        val difference = (antennas.first.row - antennas.second.row) to (antennas.first.column - antennas.second.column)
        return Point(antennas.first.row + difference.first, antennas.first.column + difference.second, "#")
    }

    fun calcLowerAntinode(antennas: Pair<Point, Point>): Point {
        val difference = (antennas.first.row - antennas.second.row) to (antennas.first.column - antennas.second.column)
        return Point(antennas.second.row - difference.first, antennas.second.column - difference.second, "#")
    }


    fun readData(input: List<String>) = input.mapIndexed { row, d ->
        d.split("").filter { it.isNotBlank() }
            .mapIndexed { col, antenna -> Point(row, col, antenna) }
    }.toList()


    data class Point(val row: Int, val column: Int, val antenna: String) {
        fun isInMatrix(rows: Int, columns: Int): Boolean {
            return row >= 0 && row < rows && column >= 0 && column < columns
        }
    }

}